Autogenerated sequences for PRIMARY KEY values (#185)
crate/cratedb-guide: docs/performance/inserts/sequences.rst

If you are coming to CrateDB from systems like PostgreSQL or MySQL one thing you
may notice is that CrateDB does not have a built-in auto-incrementing "serial"
data type. This is because CrateDB is a distributed database, designed to scale
horizontally, and this is achieved by making so that as many operations as
possible can be completed by each node independently, without having to
coordinate with other nodes. Maintaining an auto-increment value would mean that
nodes would need to check with each other before allocating a new value to a
row, creating a bottleneck and hindering our ability to achieve `extremely fast
ingestion speeds`_.

That said, there are many alternatives available and we can also implement true
consistent/synchronized sequences if we want to.

************************************
 Using a timestamp as a primary key
************************************

This option involves declaring a column as follows:

.. code:: psql

   BIGINT DEFAULT now() PRIMARY KEY

**Pros:** Always increasing number - ideal if we need to timestamp records
creation anyway

**Cons:** gaps between the numbers, not suitable if we may have more than one

> record on the same millisecond

Reviewer: In Sphinx/docutils jargon, this language construct is called [Field Lists](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#field-lists). It is traditionally used to annotate function arguments in inline documentation.
```
:field name: field body
```

> **Cons:** gaps between the numbers, not suitable if we may have more than one
> record on the same millisecond

Reviewer: ```suggestion
**Cons:** Gaps between sequence numbers. Limited to millisecond granularity.
```

*************
 Using UUIDs
*************

This option involves declaring a column as follows:

.. code:: psql

   TEXT DEFAULT gen_random_text_uuid() PRIMARY KEY

**Pros:** Globally unique, no risk of conflicts if merging things from different
tables/environments

> **Cons:** No order guarantee, not as human-friendly as numbers

Reviewer: ```suggestion
**Cons:** No order guarantee. Not as human-friendly as numbers. String format may
not be applicable to cover all scenarios.
```

************************
 Use UUIDv7 identifiers
************************

`Version 7 UUIDs`_ are a relatively new kind of UUIDs which feature a
time-ordered value. We can use these in CrateDB with an UDF_ with the code from

> https://github.com/nalgeon/uuidv7/pull/37 (update once merged)

Reviewer: Excellent. It was merged one hour ago.

Reviewer: ```suggestion
time-ordered value. We can use it in CrateDB with an UDF_ and the code from [uuidv7.cratedb].

[uuidv7.cratedb]: https://github.com/nalgeon/uuidv7/blob/main/src/uuidv7.cratedb
```

> **Pros:** Same as gen_random_text_uuid above but almost sequential

Reviewer: ```suggestion
**Pros:** Same as `gen_random_text_uuid` above but almost sequential
```

**Cons:** not as human-friendly as numbers and slight performance impact from
UDF use

*********************************
 Use IDs from an external system
*********************************

> In cases where data is flowing to CrateDB from another system, we can declare
> the column on the CrateDB side without any defaults and just ingest the
> auto-incrementing values coming from the source database.

Reviewer: ```suggestion
In cases where data is imported into CrateDB from external systems that employ
identifier governance, CrateDB does not need to generate any identifier values at
all. In this case, primary key information is inherited from the source system.
```

> See
> https://cratedb.com/blog/replicating-data-from-other-databases-to-cratedb-with-debezium-and-kafka
> for an example.

Reviewer: ```suggestion
See [Replicating data from other databases to CrateDB with Debezium and Kafka] for an example.

[Replicating data from other databases to CrateDB with Debezium and Kafka]: https://cratedb.com/blog/replicating-data-from-other-databases-to-cratedb-with-debezium-and-kafka
```

*********************
 Implement sequences
*********************

> This approach involves a table to keep the latest values that have been consumed
> and client side code to keep it up-to-date in a way that guarantees unique
> values even when many ingestion processes run in parallel.

Reviewer: ```suggestion
Implementing sequences using a manual approach involves:
- A bookkeeping table to maintain the latest values that have been consumed.
- Client-side code to keep it up-to-date in a way that guarantees unique
  values even when many ingestion processes run in parallel.
```

numbers and letters - etc)

**Cons:** Need logic for the optimistic update implemented
client-side, the sequences table becomes a bottleneck so not suitable for
high-velocity ingestion scenarios

We will first create a table to keep the latest values for our sequences:

.. code:: psql

   CREATE TABLE sequences (
           name TEXT PRIMARY KEY,
           last_value BIGINT
   ) CLUSTERED INTO 1 SHARDS;

We will then initialize it with one new sequence at 0:

.. code:: psql

   INSERT INTO sequences (name,last_value)
   VALUES ('mysequence',0);

And we are going to do an example with a new table defined as follows:

.. code:: psql

   CREATE TABLE mytable (
           id BIGINT PRIMARY KEY,
           field1 TEXT
   );

The Python code below reads the last value used from the sequences table, and
then attempts an `optimistic UPDATE`_ with a ``RETURNING`` clause, if a contending
process already consumed the identity nothing will be returned so our process
will retry until a value is returned, then it uses that value as the new ID for
the record we are inserting into the ``mytable`` table.

.. code:: python

> 

Reviewer: Then, just invoke the program per `uv run <program.py>`.

> 

Reviewer: It might not be applicable in this specific case, if it adds too much header noise.
However, you may find it helpful in other situations, so please read it merely FYI.

   import records

   db = records.Database("crate://")
   sequence_name = "mysequence"

>    while True:

Reviewer: That reads dangerous, potentially implementing a busy loop, but I guess you know what you are doing?
Still, shall we possibly add a timeout on this identifier-polling operation, so the loop may not accidentally spin forever?

       select_query = """
                           SELECT last_value,
                                   _seq_no,
                                   _primary_term
                           FROM sequences
                           WHERE name = :sequence_name;
                           """
       row = db.query(select_query, sequence_name=sequence_name).first()
       new_value = row.last_value + 1

       update_query = """
                           UPDATE sequences
                           SET last_value = :new_value
                           WHERE name = :sequence_name
                             AND _seq_no = :seq_no
                             AND _primary_term = :primary_term
                           RETURNING last_value;
                   """
       if (
           str(
               db.query(
                   update_query,
                   new_value=new_value,
                   sequence_name=sequence_name,
                   seq_no=row._seq_no,
                   primary_term=row._primary_term,
               ).all()
           )

>            != "[]"

Author: I really looked for an alternative but could not find it. There would be plenty of options giving up on the records library but I like that it makes the rest of the code so simple.

>            != "[]"

Reviewer: Thanks. It certainly should not block the merge. Because I love that you also like the `records` library, I may check this spot on a subsequent iteration to possibly find a better way to compare for empty?

Why not just do it this way, if it's semantically correct that you want to break if the result is **not** empty?
```python
result = db.query(....)
if result:
    break
```

.. _autogenerated_sequences_performance:

###########################################################
 Autogenerated sequences and PRIMARY KEY values in CrateDB
###########################################################

> If you are coming to CrateDB from systems like PostgreSQL or MySQL one thing you

Reviewer 2: I have tried to rephrase the intro/motivation a bit:

As you begin working with CrateDB, you might be puzzled why CrateDB does not have a built-in, auto-incrementing "serial" data type as PostgreSQL or MySQL.

As a distributed database , designed to scale horizontally, CrateDB needs as many operations as possible to complete independently on each node without any coordination between nodes. 

Maintaining a global auto-increment value will require that a node would need to check with all other nodes before allocating a new value - and preventing all other nodes to allocate new values. This bottleneck will be hindering our ability to achieve extremely_fast_ingestion_speeds.

coordinate with other nodes. Maintaining an auto-increment value would mean that
nodes would need to check with each other before allocating a new value to a
row, creating a bottleneck and hindering our ability to achieve `extremely fast
ingestion speeds`_.

That said, there are many alternatives available and we can also implement true
consistent/synchronized sequences if we want to.

************************************
 Using a timestamp as a primary key
************************************

This option involves declaring a column as follows:

.. code:: psql

   BIGINT DEFAULT now() PRIMARY KEY

**Pros:** Always increasing number - ideal if we need to timestamp records
creation anyway

**Cons:** gaps between the numbers, not suitable if we may have more than one
record on the same millisecond

*************
 Using UUIDs
*************

This option involves declaring a column as follows:

.. code:: psql

   TEXT DEFAULT gen_random_text_uuid() PRIMARY KEY

**Pros:** Globally unique, no risk of conflicts if merging things from different
tables/environments

**Cons:** No order guarantee. Not as human-friendly as numbers. String format may

> not be applicable to cover all scenarios.

Reviewer 3: Additionally, I think it's worth adding that no range queries are possible, as UUIDs aren't sequential. This is one of the main benefits of algorithms like a [snowflake_id](https://en.wikipedia.org/wiki/Snowflake_ID)

.. _autogenerated_sequences_performance:

###########################################################
 Autogenerated sequences and PRIMARY KEY values in CrateDB
###########################################################

As you begin working with CrateDB, you might be puzzled why CrateDB does not
have a built-in, auto-incrementing "serial" data type as PostgreSQL or MySQL.

As a distributed database, designed to scale horizontally, CrateDB needs as
many operations as possible to complete independently on each node without
any coordination between nodes.

Maintaining a global auto-increment value requires that a node checks with 
other nodes before allocating a new value. This bottleneck would be hindering
our ability to achieve `extremely fast ingestion speeds`_.

That said, there are many alternatives available and we can also implement true
consistent/synchronized sequences if we want to.

************************************
 Using a timestamp as a primary key
************************************

This option involves declaring a column as follows:

.. code:: psql

   BIGINT DEFAULT now() PRIMARY KEY

> :Pros: Always increasing number - ideal if we need to timestamp records
> creation anyway
> 
> 
> :Cons: gaps between the numbers, not suitable if we may have more than one
> record on the same millisecond

Reviewer: This snippet causes the _WARNING: Field list ends without a blank line; unexpected unindent._

-- https://github.com/crate/cratedb-guide/actions/runs/14473917871/job/40594700368?pr=185#step:4:581

I think this might work better:
```suggestion
:Pros: Always increasing number - ideal if we need to timestamp records
       creation anyway
:Cons: gaps between the numbers, not suitable if we may have more than one
       record on the same millisecond
```

.. code:: psql

   TEXT DEFAULT gen_random_text_uuid() PRIMARY KEY

:Pros: Globally unique, no risk of conflicts if merging things from different
tables/environments

:Cons: No order guarantee. Not as human-friendly as numbers. String format may
not be applicable to cover all scenarios. Range queries are not possible.

************************
 Use UUIDv7 identifiers
************************

`Version 7 UUIDs`_ are a relatively new kind of UUIDs which feature a
time-ordered value. We can use these in CrateDB with an UDF_ with the code from
`UUIDv7 in N languages`_.

:Pros: Same as `gen_random_text_uuid` above but almost sequential, which 
enables range queries.

:Cons: not as human-friendly as numbers and slight performance impact from
UDF use

*********************************
 Use IDs from an external system
*********************************

In cases where data is imported into CrateDB from external systems that employ
identifier governance, CrateDB does not need to generate any identifier values
and primary key values can be inserted as-is from the source system.

See [Replicating data from other databases to CrateDB with Debezium and Kafka] for an example.

> [Replicating data from other databases to CrateDB with Debezium and Kafka]: https://cratedb.com/blog/replicating-data-from-other-databases-to-cratedb-with-debezium-and-kafka

Reviewer: That's Markdown. ;]

NB: We will convert this file and others to Markdown anyway, to provide better support for [llms.txt](https://github.com/crate/tech-content/issues/135), so it does not necessarily needs to be fixed.
